11_FFTBloom/app_resources/image_ifft_first_axis.hlsl

```diff
@@ -0,0 +1,159 @@
+		// want it to hold conj(C1) + i*conj(C2). So we first do conj(C1 + iC2) to yield conj(C1) - i*conj(C2). Then we subtract conj(C1) to get -i*conj(C2), negate that to get i * conj(C2), and finally
+		// add conj(C1) back to have conj(C1) + i * conj(C2).
+		for (uint32_t localElementIndex = 1; localElementIndex < ELEMENTS_PER_THREAD; localElementIndex += 2)
+		{
+			// Thread 0's first odd element is Nyquist, which was packed alongside Zero - this means that what was said above breaks in this particular case and needs special treatment
+			if (!workgroup::SubgroupContiguousIndex() && 1 == localElementIndex)
+			{
+				// preloaded[1] currently holds trash - this is because 0 and Nyquist are the only fixed points of T -> -T. 
+				// preloaded[0] currently holds (C1(Z) - C2(N)) + i * (C1(N) + C2(Z)). This is because of how we loaded the even elements of both columns.
+				// We want preloaded[0] to hold C1(Z) + i * C2(Z) and preloaded[1] to hold C1(N) + i * C2(N).
+				// We can re-load C2(Z) + i * C2(N) and use it to unpack the values
+				complex_t<scalar_t> c2 = rowMajorAccessor.get(rowMajorOffset(2 * glsl::gl_WorkGroupID().x + 1, 0));
+				complex_t<scalar_t> p1 = { preloaded[0].imag() - c2.real(), c2.imag() };
+				preloaded[1] = p1;
+				complex_t<scalar_t> p0 = { preloaded[0].real() + c2.imag() , c2.real()};
+				preloaded[0] = p0;
+			}
+			else 
+			{
+				complex_t<scalar_t> otherThreadEven = conj(getDFTMirror<SharedmemAdaptor>(localElementIndex, sharedmemAdaptor));
+				otherThreadEven = otherThreadEven - preloaded[localElementIndex];
+				otherThreadEven = otherThreadEven * scalar_t(-1);
+				preloaded[localElementIndex] = preloaded[localElementIndex] + otherThreadEven;
+			}
+		}
+	}
+
+	void unload(uint32_t channel)
+	{
+		uint32_t2 imageDimensions;
+		convolvedImage.GetDimensions(imageDimensions.x, imageDimensions.y);
+		const uint32_t padding = uint32_t(FFT_LENGTH - imageDimensions.y) >> 1;
+		for (uint32_t localElementIndex = 0; localElementIndex < ELEMENTS_PER_THREAD; localElementIndex++)
+		{
+			const uint32_t index = _NBL_HLSL_WORKGROUP_SIZE_ * localElementIndex | workgroup::SubgroupContiguousIndex();
+			const uint32_t paddedIndex = index - padding;
+			if (paddedIndex >= 0 && paddedIndex < imageDimensions.y)
+			{
+				vector<scalar_t, 4> texValue = convolvedImage.Load(uint32_t2(2 * glsl::gl_WorkGroupID().x, paddedIndex));
+				texValue[channel] = scalar_t(preloaded[localElementIndex].real());
+				texValue.a = scalar_t(1);
+				convolvedImage[uint32_t2(2 * glsl::gl_WorkGroupID().x, paddedIndex)] = texValue;
+
+				texValue = convolvedImage.Load(uint32_t2(2 * glsl::gl_WorkGroupID().x + 1, paddedIndex));
+				texValue[channel] = scalar_t(preloaded[localElementIndex].imag());
+				texValue.a = scalar_t(1);
+				convolvedImage[uint32_t2(2 * glsl::gl_WorkGroupID().x + 1, paddedIndex)] = texValue;
```

Comment:
this is probably why you want to preload all channels at once, and bite the occupancy bullet

Reply:
![image](https://github.com/user-attachments/assets/dc47c44f-00c8-41bb-8cfb-6636f54c8b6e)
I have a lot of unallocated warps, wouldn't this make it worse? Or is the fetch time THAT impactful? Once I'm done cleaning up the easier stuff I'll change it do do an all-channel preload and compare profiles

Reply:
its worth investigating, you know the unallocated warps only become an issue if there's actually work they could perform.

Texture reads have large latencies (like 1000 cycles), so reading 3 times redundantly could have a worse impact than higher register pressure, but if in doubt then profile.